pkg/dns/plugin.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	dnsapi "istio.io/istio/pkg/dns/coredns/pb"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+
+	"google.golang.org/grpc"
+)
+
+// Based on istio-ecosystem/istio-coredns-plugin
+// Changes from original:
+// - runs inside istiod, using Istio main gRPC server
+// - instead of directly reading from K8S, use istiod store.
+// - removed "log" - switching to istio log.
+// - refactored Query, so both DNS native interface and coredns grpc plugin are implemented
+// - added parts of istio-ecosystem/dns-discovery, to provide in process DNS
+
+// K8S DNS uses container port 10053, and appears to have a dnsmasq container
+// alongside.
+
+// IstioServiceEntries exposes a DNS interface to internal Istiod service database.
+// This can be used:
+// - as a CoreDNS gRPC plugin
+// - as a DNS-over-TLS resolver, with support for forwarding to k8s or upstream
+// - for debug - a plain DNS-over-UDP interface.
+//
+// The code is currently targeted for Istiod, with a per/pod or per/VM coreDNS
+// forwarding to it, and using the same XDS grpc server and cert.
+//
+// In future we may embed it in istio-agent as well, using XDS to sync the config store.
+type IstioServiceEntries struct {
+	configStore model.IstioConfigStore
+	mapMutex    sync.RWMutex
+	mux         *dns.ServeMux
+	dnsEntries  map[string][]net.IP
+	client      *dns.Client
+	server      *dns.Server
+}
+
+var (
+	// DNSAddr is the env controlling the DNS-over-TLS server.
+	// By default will be active, set to empty string to disable DNS functionality.
+	DNSAddr = env.RegisterStringVar("dnsAddr", ":15053", "DNS listen address")
+)
+
+// Initialize the CoreDNS gRPC plugin.
+// vip is the default IP to assign to entries that are not found.
+func InitCoreDNS(grpcServer *grpc.Server, vip string, cfgStore model.IstioConfigStore) *IstioServiceEntries {
```

Comment:
Yes, that's the plan. Plus probably moving the struct into PushContext - and extending it to also serve cluster IPs from the cache instead of forwarding to k8s.